lib/spack/spack/spec.py

```diff
@@ -1829,6 +1807,43 @@ def concretize(self):
             if s.namespace is None:
                 s.namespace = spack.repo.repo_for_pkg(s.name).namespace
 
+            # Add any patches from the package to the spec.
+            patches = []
+            for cond, patch_list in s.package_class.patches.items():
+                if s.satisfies(cond):
+                    for patch in patch_list:
+                        patches.append(patch.sha256)
+            if patches:
+                # Special-case: keeps variant values unique but ordered.
+                s.variants['patches'] = MultiValuedVariant('patches', ())
+                mvar = s.variants['patches']
+                mvar._value = mvar._original_value = tuple(dedupe(patches))
```

Comment:
Here we are using a MV variant with a reserved name to keep track of patches, and correctly set it after concretization. I am curious why we didn't go with an additional `Spec` attribute and a modification to `Spec.to_node_dict`.